lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java

```diff
@@ -122,14 +123,19 @@ public static Predicate buildPredicate(
         return new TupleDomainParquetPredicate(parquetTupleDomain, columnReferences.build(), timeZone);
     }
 
-    public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData block, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain)
+    public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData block, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain, ColumnIndexStore ciStore, boolean readColumnIndex)
             throws ParquetCorruptionException
     {
         Map<ColumnDescriptor, Statistics<?>> columnStatistics = getStatistics(block, descriptorsByPath);
         if (!parquetPredicate.matches(block.getRowCount(), columnStatistics, dataSource.getId())) {
             return false;
         }
 
+        // Page stats is finer grained but relatively more expensive, so we do the filtering after above block filtering.
+        if (ciStore != null && readColumnIndex && !parquetPredicate.matches(block.getRowCount(), ciStore, dataSource.getId())) {
+            return false;
+        }
```

Comment:
With the `Optional` argument, this becomes
```java
if (indexStore.isPresent()
        && !parquetPredicate.matches(block.getRowCount(), indexStore.get(), dataSource.getId())) {
```
Unfortunately, the exception means a proper optional chain doesn't work.